File: src/com/intellij/idea/plugin/hybris/impex/psi/references/TSItemReference.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.impex.psi.references;

import com.intellij.idea.plugin.hybris.impex.psi.ImpexHeaderTypeName;
import com.intellij.idea.plugin.hybris.psi.reference.TSReferenceBase;
import com.intellij.idea.plugin.hybris.psi.utils.PsiUtils;
import com.intellij.idea.plugin.hybris.system.type.psi.reference.result.EnumResolveResult;
import com.intellij.idea.plugin.hybris.system.type.psi.reference.result.ItemResolveResult;
import com.intellij.idea.plugin.hybris.system.type.psi.reference.result.RelationResolveResult;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.psi.ResolveResult;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

/**
 * Created by Martin Zdarsky-Jones (martin.zdarsky@hybris.com) on 15/06/2016.
 */
class TSItemReference extends TSReferenceBase<ImpexHeaderTypeName> {

    public TSItemReference(@NotNull final ImpexHeaderTypeName owner) {
        super(owner);
    }

    @NotNull
    @Override
    public ResolveResult[] multiResolve(final boolean incompleteCode) {
        final ProgressIndicator indicator = ProgressManager.getInstance().getProgressIndicator();
        if (indicator != null && indicator.isCanceled()) {
            return ResolveResult.EMPTY_ARRAY;
        }

        final var cachedResolveResult = getElement().getUserData(ImpexHeaderTypeNameMixin.CACHE_KEY);
        if (cachedResolveResult != null) {
            return PsiUtils.getValidResults(cachedResolveResult);
        }

        final var lookingForName = getValue();
        final var metaService = getMetaModelAccess();

        final var resolvedResults = Optional.ofNullable(metaService.findMetaItemByName(lookingForName))
                                            .map(it -> it.getDeclarations().stream()
                                                         .map(ItemResolveResult::new)
                                                         .toArray(ResolveResult[]::new))
                                            .or(() -> Optional.ofNullable(metaService.findMetaEnumByName(lookingForName))
                                                              .map(it -> new ResolveResult[]{new EnumResolveResult(it)}))
                                            .or(() -> Optional.ofNullable(metaService.findMetaRelationByName(lookingForName))
                                                              .map(it -> new ResolveResult[]{new RelationResolveResult(it)}))
                                            .orElse(ResolveResult.EMPTY_ARRAY);
        getElement().putUserData(ImpexHeaderTypeNameMixin.CACHE_KEY, resolvedResults);
        return PsiUtils.getValidResults(resolvedResults);
    }

}

File: src/com/intellij/idea/plugin/hybris/system/cockpitng/psi/TSItemReference.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.system.cockpitng.psi

import com.intellij.codeInsight.highlighting.HighlightedReference
import com.intellij.idea.plugin.hybris.common.HybrisConstants
import com.intellij.idea.plugin.hybris.impex.psi.references.result.EnumResolveResult
import com.intellij.idea.plugin.hybris.impex.psi.references.result.ItemResolveResult
import com.intellij.idea.plugin.hybris.impex.psi.references.result.RelationResolveResult
import com.intellij.idea.plugin.hybris.psi.reference.TSReferenceBase
import com.intellij.idea.plugin.hybris.system.type.meta.model.TSGlobalMetaEnum
import com.intellij.idea.plugin.hybris.system.type.meta.model.TSGlobalMetaItem
import com.intellij.idea.plugin.hybris.system.type.meta.model.TSGlobalMetaRelation
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.ResolveResult

/**
 * https://help.sap.com/docs/SAP_COMMERCE/5c9ea0c629214e42b727bf08800d8dfa/8b59d395866910149db8889c5087a5ef.html?locale=en-US&q=ExtendedMultiReference
 *
 * see, standard-editors-spring.xml
 */
class TSItemReference(element: PsiElement) : TSReferenceBase<PsiElement>(element), PsiPolyVariantReference, HighlightedReference {

    override fun calculateDefaultRangeInElement(): TextRange {
        val text = element.text.trim()
            .replace("\"", "")

        return regexes
            .firstOrNull { it.matches(text) }
            ?.let { _ ->
                val offset = element.text.indexOfLast { it == '(' } + 1
                val length = element.text.indexOfFirst { it == ')' } - offset
                TextRange.from(offset, length)
            }
            ?: TextRange.from(1, element.textLength - HybrisConstants.QUOTE_LENGTH)
    }

    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
        val lookingForName = value

        return metaModelAccess.findMetaItemByName(lookingForName)
            ?.let { resolve(it) }
            ?: metaModelAccess.findMetaEnumByName(lookingForName)
                ?.let { resolve(it) }
            ?: metaModelAccess.findMetaRelationByName(lookingForName)
                ?.let { resolve(it) }
            ?: emptyArray()
    }

    private fun resolve(meta: TSGlobalMetaItem): Array<ResolveResult> = meta.retrieveAllDoms()
        .map { ItemResolveResult(it) }
        .toTypedArray()

    private fun resolve(meta: TSGlobalMetaEnum): Array<ResolveResult> = meta.retrieveAllDoms()
        .map { EnumResolveResult(it) }
        .toTypedArray()

    private fun resolve(meta: TSGlobalMetaRelation): Array<ResolveResult> = meta.retrieveAllDoms()
        .map { RelationResolveResult(it) }
        .toTypedArray()

    companion object {
        private val regexes = arrayOf(
            "^Localized\\((.*)\\)\$".toRegex(),
            "^LocalizedSimple\\((.*)\\)\$".toRegex(),
            "^List\\((.*)\\)\$".toRegex(),
            "^Reference\\((.*)\\)\$".toRegex(),
            "^FixedValuesReference\\((.*)\\)\$".toRegex(),
            "^MultiReference-(COLLECTION|LIST|SET)\\((.*)\\)\$".toRegex(),
            "^ExtendedMultiReference-(COLLECTION|LIST|SET)\\((.*)\\)\$".toRegex(),
            "^EnumMultiReference-(COLLECTION|LIST|SET)\\((.*)\\)\$".toRegex(),
        )
    }

}
